File: ios/MitiMaiti/ViewModels/ChatViewModel.swift
import SwiftUI
import Combine

@MainActor
class ChatViewModel: ObservableObject {
    @Published var messages: [Message] = []
    @Published var messageText = ""
    @Published var isLoading = false
    @Published var isSending = false
    @Published var isOtherTyping = false
    @Published var match: Match?
    @Published var error: String?
    @Published var chatUnlocked = false

    private let api = APIService.shared
    private let currentUserId = "current-user-id"

    /// Injected so that when a reply unlocks the match we can update the
    /// shared InboxViewModel and move the match into the Chats section.
    weak var inboxViewModel: InboxViewModel?

    var hasMessages: Bool { !messages.isEmpty }

    // MARK: - Respect-First Lock State

    /// Whether the chat is in "locked" state:
    /// - The current user sent the first message
    /// - The other user has NOT replied yet
    /// - Current user CANNOT send another message until they reply
    var isLockedForMe: Bool {
        guard let match else { return false }
        // If match says locked AND I'm the one who sent first → I'm blocked
        return match.firstMsgLocked && match.firstMsgBy == currentUserId
    }

    /// Whether the other user needs to send the first message
    /// (match exists but nobody has messaged yet)
    var awaitingFirstMessage: Bool {
        guard let match else { return false }
        return !match.hasFirstMessage
    }

    /// Whether the countdown timer should show
    var showCountdown: Bool {
        match?.showCountdown ?? false
    }

    /// Dynamic banner message based on lock state
    var lockBannerMessage: (title: String, subtitle: String)? {
        guard let match else { return nil }

        if !match.hasFirstMessage {
            // No one has messaged yet
            return ("Send the first message!", "Break the ice before the timer runs out")
        }

        if match.firstMsgLocked && match.firstMsgBy == currentUserId {
            // I sent first, waiting for their reply
            return ("Waiting for reply...", "Your message has been sent. You can send another message once \(match.otherUser.displayName) replies.")
        }

        if match.firstMsgLocked && match.firstMsgBy != currentUserId {
            // They sent first, I need to reply
            return ("\(match.otherUser.displayName) sent the first message!", "Reply to keep the conversation going")
        }

        return nil
    }

    /// Whether the text input should be disabled
    var inputDisabled: Bool {
        isLockedForMe
    }

    /// Placeholder text for the input field
    var inputPlaceholder: String {
        if isLockedForMe {
            return "Waiting for \(match?.otherUser.displayName ?? "them") to reply..."
        }
        if awaitingFirstMessage {
            return "Send the first message..."
        }
        return "Type a message..."
    }

    // MARK: - Load Messages

    func loadMessages(for match: Match) {
        self.match = match

        // Serve from cache immediately — avoids blank screen on re-navigation
        let cached = MessageRepository.shared.getMessages(matchId: match.id)
        if !cached.isEmpty {
            messages = cached
            checkAndUnlockIfReplied()
            return
        }

        isLoading = true

        Task {
            do {
                let msgs = try await api.fetchMessages(matchId: match.id)
                messages = msgs.sorted { $0.createdAt < $1.createdAt }
                MessageRepository.shared.setMessages(matchId: match.id, msgs: messages)
                isLoading = false

                // Check if the other user has already replied (unlock state)
                checkAndUnlockIfReplied()
            } catch {
                self.error = error.localizedDescription
                isLoading = false
            }
        }
    }

    // MARK: - Send Message

    func sendMessage() {
        // If we're editing, route to saveEdit instead
        if editingMessageId != nil {
            saveEdit()
            return
        }
        let text = messageText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, let matchId = match?.id else { return }

        // Respect-First enforcement: block if locked for me
        if isLockedForMe {
            error = "Please wait for \(match?.otherUser.displayName ?? "them") to reply to your first message."
            return
        }

        let tempMsg = Message(
            matchId: matchId,
            senderId: currentUserId,
            content: text,
            status: .sending
        )
        messages.append(tempMsg)
        messageText = ""
        isSending = true

        // If this is the first message in the match, engage the lock
        let isFirstMessage = !match!.hasFirstMessage

        Task {
            do {
                let sent = try await api.sendMessage(matchId: matchId, content: text)
                if let idx = messages.firstIndex(where: { $0.id == tempMsg.id }) {
                    messages[idx] = sent
                }
                MessageRepository.shared.setMessages(matchId: matchId, msgs: messages)
                isSending = false

                if isFirstMessage {
                    // Lock engaged: I sent first, now waiting for their reply
                    match?.firstMsgBy = currentUserId
                    match?.firstMsgLocked = true
                    match?.firstMsgAt = Date()
                }

                // In demo mode: simulate a reply from the other user
                simulateReply(matchId: matchId)
            } catch {
                self.error = error.localizedDescription
                isSending = false
            }
        }
    }

    @Published var editingMessageId: String? = nil

    func startEdit(_ message: Message) {
        guard message.msgType == .text else { return }
        editingMessageId = message.id
        var text = message.content
        if text.hasSuffix(" [edited]") {
            text = String(text.dropLast(" [edited]".count))
        } else if text.hasSuffix("[edited]") {
            text = String(text.dropLast("[edited]".count))
        }
        messageText = text.trimmingCharacters(in: .whitespaces)
    }

    func cancelEdit() {
        editingMessageId = nil
        messageText = ""
    }

    func saveEdit() {
        guard let id = editingMessageId, let matchId = match?.id else { return }
        let trimmed = messageText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        let withMarker = trimmed.contains("[edited]") ? trimmed : "\(trimmed) [edited]"
        if let idx = messages.firstIndex(where: { $0.id == id }) {
            messages[idx].content = withMarker
            MessageRepository.shared.setMessages(matchId: matchId, msgs: messages)
        }
        editingMessageId = nil
        messageText = ""
    }

    func toggleReaction(_ message: Message, emoji: String) {
        guard Message.allowedReactions.contains(emoji) else { return }
        guard let matchId = match?.id else { return }
        if let idx = messages.firstIndex(where: { $0.id == message.id }) {
            messages[idx].reaction = (messages[idx].reaction == emoji) ? nil : emoji
            MessageRepository.shared.setMessages(matchId: matchId, msgs: messages)
        }
    }

    func deleteMessage(_ message: Message) {
        guard let matchId = match?.id else { return }
        messages.removeAll { $0.id == message.id }
        MessageRepository.shared.setMessages(matchId: matchId, msgs: messages)
        if editingMessageId == message.id {
            cancelEdit()
        }
    }

    func sendVoice(localUrl: String, durationSeconds: Int) {
        guard let matchId = match?.id else { return }
        if isLockedForMe {
            error = "Please wait for \(match?.otherUser.displayName ?? "them") to reply to your first message."
            return
        }

        let isFirstMessage = !match!.hasFirstMessage

        let msg = Message(
            matchId: matchId,
            senderId: currentUserId,
            content: "",
            mediaUrl: localUrl,
            msgType: .voice,
            status: .sent,
            durationSeconds: durationSeconds
        )
        messages.append(msg)
        MessageRepository.shared.setMessages(matchId: matchId, msgs: messages)

        if isFirstMessage {
            match?.firstMsgBy = currentUserId
            match?.firstMsgLocked = true
            match?.firstMsgAt = Date()
        }

        simulateReply(matchId: matchId)
    }

    func sendImage(localUrl: String) {
        guard let matchId = match?.id else { return }
        if isLockedForMe {
            error = "Please wait for \(match?.otherUser.displayName ?? "them") to reply to your first message."
            return
        }

        let isFirstMessage = !match!.hasFirstMessage

        let msg = Message(
            matchId: matchId,
            senderId: currentUserId,
            content: "",
            mediaUrl: localUrl,
            msgType: .photo,
            status: .sent
        )
        messages.append(msg)
        MessageRepository.shared.setMessages(matchId: matchId, msgs: messages)

        if isFirstMessage {
            match?.firstMsgBy = currentUserId
            match?.firstMsgLocked = true
            match?.firstMsgAt = Date()
        }

        simulateReply(matchId: matchId)
    }

    /// Real-backend photo upload (added on the Mac side via APIService.sendChatMedia).
    func sendPhoto(imageData: Data) {
        guard let matchId = match?.id else { return }
        if isLockedForMe {
            error = "Please wait for \(match?.otherUser.displayName ?? "them") to reply to your first message."
            return
        }
        let tempMsg = Message(matchId: matchId, senderId: currentUserId,
                              content: "Uploading...", msgType: .photo, status: .sending)
        messages.append(tempMsg)
        let isFirstMessage = !match!.hasFirstMessage
        isSending = true

        Task {
            do {
                let sent = try await api.sendChatMedia(matchId: matchId, imageData: imageData)
                if let idx = messages.firstIndex(where: { $0.id == tempMsg.id }) {
                    messages[idx] = sent
                }
                MessageRepository.shared.setMessages(matchId: matchId, msgs: messages)
                isSending = false
                if isFirstMessage {
                    match?.firstMsgBy = currentUserId
                    match?.firstMsgLocked = true
                    match?.firstMsgAt = Date()
                }
            } catch {
                self.error = error.localizedDescription
                if let idx = messages.firstIndex(where: { $0.id == tempMsg.id }) {
                    messages.remove(at: idx)
                }
                isSending = false
            }
        }
    }

    func sendIcebreaker(_ question: String) {
        guard let matchId = match?.id else { return }

        // Icebreakers also count as first message
        if isLockedForMe {
            error = "Please wait for \(match?.otherUser.displayName ?? "them") to reply first."
            return
        }

        let isFirstMessage = !match!.hasFirstMessage

        let msg = Message(
            matchId: matchId,
            senderId: currentUserId,
            content: question,
            msgType: .icebreaker,
            status: .sending
        )
        messages.append(msg)

        Task {
            do {
                let sent = try await api.sendMessage(matchId: matchId, content: question, type: .icebreaker)
                if let idx = messages.firstIndex(where: { $0.id == msg.id }) {
                    messages[idx] = sent
                }
                MessageRepository.shared.setMessages(matchId: matchId, msgs: messages)

                if isFirstMessage {
                    match?.firstMsgBy = currentUserId
                    match?.firstMsgLocked = true
                    match?.firstMsgAt = Date()
                }

                simulateReply(matchId: matchId)
            } catch {
                self.error = error.localizedDescription
            }
        }
    }

    // MARK: - Receive Message (from other user)

    /// Called when a message arrives from the other user (via socket or simulation)
    func receiveMessage(_ message: Message) {
        messages.append(message)
        if let matchId = match?.id {
            MessageRepository.shared.addMessage(matchId: matchId, message: message)
        }

        // Trigger a message notification
        if let match {
            NotificationManager.shared.addNotification(
                type: .message,
                title: "New message from \(match.otherUser.displayName)",
                body: message.content,
                actionData: match.id
            )
        }

        // If the match was locked and the other user just replied → UNLOCK
        if let match, match.firstMsgLocked {
            // The other user has replied — unlock the conversation
            self.match?.firstMsgLocked = false
            self.chatUnlocked = true

            // Match is now fully active: no expiry, promote status
            self.match?.expiresAt = nil
            self.match?.status = .active

            // Propagate to InboxViewModel so MatchesView moves the row
            // from the timer-avatar section into the permanent Chats list.
            inboxViewModel?.activateMatch(id: match.id)
        }
    }

    // MARK: - Check & Unlock

    /// After loading messages, check if the other user has already replied
    private func checkAndUnlockIfReplied() {
        guard let match, match.firstMsgLocked, match.firstMsgBy == currentUserId else { return }

        let otherUserId = match.otherUser.id
        let hasReply = messages.contains { $0.senderId == otherUserId }

        if hasReply {
            // Other user has already replied — unlock
            self.match?.firstMsgLocked = false
            self.chatUnlocked = true
            self.match?.expiresAt = nil
            self.match?.status = .active

            // Propagate to InboxViewModel so MatchesView reflects the change
            inboxViewModel?.activateMatch(id: match.id)
        }
    }

    // MARK: - Mark Messages As Read

    /// Marks all unread messages from the other user as "read"
    func markUnreadMessagesAsRead() {
        guard let match else { return }
        let otherUserId = match.otherUser.id
        for i in messages.indices {
            if messages[i].senderId == otherUserId &&
               (messages[i].status == .sent || messages[i].status == .delivered) {
                messages[i].status = .read
            }
        }
    }

    // MARK: - Simulate Reply (Demo Mode)

    private func simulateReply(matchId: String) {
        let otherUserId = match?.otherUser.id ?? "other"

        Task {
            // Show typing after 1-2 seconds
            try? await Task.sleep(nanoseconds: UInt64.random(in: 1_000_000_000...2_000_000_000))
            isOtherTyping = true

            // Send reply after 2-4 seconds
            try? await Task.sleep(nanoseconds: UInt64.random(in: 2_000_000_000...4_000_000_000))
            isOtherTyping = false

            let replies = [
                "That's so sweet! Tell me more about yourself",
                "Haha I love that! What else do you enjoy?",
                "Same here! We have so much in common",
                "That's really interesting. I'd love to hear more!",
                "You seem like such a great person!",
                "I was thinking the same thing!",
                "Sai bhaani! So glad we matched!",
                "My family would love to hear about this!"
            ]

            let reply = Message(
                matchId: matchId,
                senderId: otherUserId,
                content: replies.randomElement()!,
                status: .delivered
            )

            // Use receiveMessage to properly handle unlock
            receiveMessage(reply)
        }
    }
}

File: android/app/src/main/java/com/mitimaiti/app/viewmodels/ChatViewModel.kt
package com.mitimaiti.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mitimaiti.app.models.*
import com.mitimaiti.app.services.APIService
import com.mitimaiti.app.services.MessageRepository
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.UUID

class ChatViewModel : ViewModel() {
    private val _messages = MutableStateFlow<List<Message>>(emptyList())
    val messages: StateFlow<List<Message>> = _messages.asStateFlow()
    private val _messageText = MutableStateFlow("")
    val messageText: StateFlow<String> = _messageText.asStateFlow()
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    private val _isSending = MutableStateFlow(false)
    val isSending: StateFlow<Boolean> = _isSending.asStateFlow()
    private val _isOtherTyping = MutableStateFlow(false)
    val isOtherTyping: StateFlow<Boolean> = _isOtherTyping.asStateFlow()
    private val _match = MutableStateFlow<Match?>(null)
    val match: StateFlow<Match?> = _match.asStateFlow()
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    private val _chatUnlocked = MutableStateFlow(false)
    val chatUnlocked: StateFlow<Boolean> = _chatUnlocked.asStateFlow()

    /**
     * Callback to notify InboxViewModel when a match becomes active (reply received).
     * Called with (matchId, lastMessage).
     */
    var onMatchActivated: ((String, String) -> Unit)? = null

    fun updateMessageText(value: String) { _messageText.value = value }
    val isLockedForMe: Boolean get() = _match.value?.let { it.firstMsgLocked && it.firstMsgBy == "current-user-id" } ?: false
    val awaitingFirstMessage: Boolean get() = _match.value?.let { it.firstMsgBy == null } ?: true
    val showCountdown: Boolean get() = _match.value?.showCountdown ?: false
    val inputDisabled: Boolean get() = isLockedForMe
    val inputPlaceholder: String get() = when { isLockedForMe -> "Waiting for reply..."; awaitingFirstMessage -> "Send the first message!"; else -> "Type a message..." }
    val callsUnlocked: Boolean get() = _match.value?.callsUnlocked ?: false

    data class LockBanner(val title: String, val subtitle: String, val isLocked: Boolean)
    val lockBannerMessage: LockBanner? get() {
        val m = _match.value ?: return null
        return when {
            isLockedForMe -> LockBanner("Message sent!", "Waiting for ${m.otherUser.displayName} to reply before you can send another", true)
            awaitingFirstMessage -> LockBanner("New Match!", "Send the first message to start the conversation", false)
            m.showCountdown && !m.callsUnlocked -> LockBanner("Keep chatting!", "Exchange messages to unlock calls & video", false)
            else -> null
        }
    }

    fun loadMessages(match: Match) {
        _match.value = match
        // Check MessageRepository first (persists across navigation)
        val cached = MessageRepository.getMessages(match.id)
        if (cached.isNotEmpty()) {
            _messages.value = cached.sortedBy { it.createdAt }
            _isLoading.value = false
            checkAndUnlockIfReplied()
            return
        }
        // Fetch from API/mock
        viewModelScope.launch {
            _isLoading.value = true
            APIService.fetchMessages(match.id)
                .onSuccess {
                    _messages.value = it.sortedBy { m -> m.createdAt }
                    // Save to repository
                    MessageRepository.setMessages(match.id, _messages.value)
                    checkAndUnlockIfReplied()
                }
                .onFailure { _error.value = "Failed to load messages" }
            _isLoading.value = false
        }
    }

    fun sendChatPhoto(bytes: ByteArray) {
        val currentMatch = _match.value ?: return
        if (isLockedForMe) return
        viewModelScope.launch {
            _isSending.value = true
            APIService.sendChatMedia(currentMatch.id, bytes).onSuccess { msg ->
                _messages.value = _messages.value + msg
                MessageRepository.addMessage(currentMatch.id, msg)
                if (currentMatch.firstMsgBy == null) {
                    _match.value = currentMatch.copy(
                        firstMsgBy = "current-user-id",
                        firstMsgLocked = true,
                        firstMsgAt = System.currentTimeMillis()
                    )
                }
            }.onFailure { _error.value = "Photo upload failed" }
            _isSending.value = false
        }
    }

    fun sendMessage() {
        // If we're editing, route to saveEdit instead of sending a new message
        if (_editingMessageId.value != null) {
            saveEdit()
            return
        }
        val text = _messageText.value.trim()
        if (text.isEmpty() || isLockedForMe) return
        val currentMatch = _match.value ?: return

        viewModelScope.launch {
            _isSending.value = true
            _messageText.value = ""

            val newMsg = Message(
                matchId = currentMatch.id,
                senderId = "current-user-id",
                content = text,
                status = MessageStatus.SENT
            )
            _messages.value = _messages.value + newMsg
            MessageRepository.addMessage(currentMatch.id, newMsg)

            // Mark as first message if none sent yet
            if (currentMatch.firstMsgBy == null) {
                _match.value = currentMatch.copy(
                    firstMsgBy = "current-user-id",
                    firstMsgLocked = true,
                    firstMsgAt = System.currentTimeMillis()
                )
            }

            APIService.sendMessage(currentMatch.id, text)
            _isSending.value = false

            // Simulate a reply after delay
            simulateReply(currentMatch.id)
        }
    }

    fun sendIcebreaker(question: String) {
        _messageText.value = question
        sendMessage()
    }

    private val _editingMessageId = MutableStateFlow<String?>(null)
    val editingMessageId: StateFlow<String?> = _editingMessageId.asStateFlow()

    fun startEdit(message: Message) {
        if (message.msgType != MessageType.TEXT) return
        _editingMessageId.value = message.id
        _messageText.value = message.content.removeSuffix(" [edited]").trimEnd()
    }

    fun cancelEdit() {
        _editingMessageId.value = null
        _messageText.value = ""
    }

    fun saveEdit() {
        val id = _editingMessageId.value ?: return
        val newText = _messageText.value.trim()
        if (newText.isEmpty()) return
        val withMarker = if (newText.endsWith("[edited]")) newText else "$newText [edited]"
        _messages.value = _messages.value.map { m ->
            if (m.id == id) m.copy(content = withMarker) else m
        }
        // persist
        _match.value?.id?.let { mid ->
            MessageRepository.setMessages(mid, _messages.value)
        }
        _editingMessageId.value = null
        _messageText.value = ""
    }

    fun toggleReaction(message: Message, emoji: String) {
        if (!Message.ALLOWED_REACTIONS.contains(emoji)) return
        val mid = _match.value?.id ?: return
        _messages.value = _messages.value.map { m ->
            if (m.id == message.id) {
                m.copy(reaction = if (m.reaction == emoji) null else emoji)
            } else m
        }
        MessageRepository.setMessages(mid, _messages.value)
    }

    fun deleteMessage(message: Message) {
        _messages.value = _messages.value.filterNot { it.id == message.id }
        _match.value?.id?.let { mid ->
            MessageRepository.setMessages(mid, _messages.value)
        }
        if (_editingMessageId.value == message.id) {
            cancelEdit()
        }
    }

    fun sendVoice(mediaUrl: String, durationSeconds: Int) {
        if (isLockedForMe) return
        val currentMatch = _match.value ?: return

        viewModelScope.launch {
            _isSending.value = true

            val newMsg = Message(
                matchId = currentMatch.id,
                senderId = "current-user-id",
                content = "",
                mediaUrl = mediaUrl,
                msgType = MessageType.VOICE,
                status = MessageStatus.SENT,
                durationSeconds = durationSeconds
            )
            _messages.value = _messages.value + newMsg
            MessageRepository.addMessage(currentMatch.id, newMsg)

            if (currentMatch.firstMsgBy == null) {
                _match.value = currentMatch.copy(
                    firstMsgBy = "current-user-id",
                    firstMsgLocked = true,
                    firstMsgAt = System.currentTimeMillis()
                )
            }

            _isSending.value = false
            simulateReply(currentMatch.id)
        }
    }

    fun sendImage(mediaUrl: String) {
        if (isLockedForMe) return
        val currentMatch = _match.value ?: return

        viewModelScope.launch {
            _isSending.value = true

            val newMsg = Message(
                matchId = currentMatch.id,
                senderId = "current-user-id",
                content = "",
                mediaUrl = mediaUrl,
                msgType = MessageType.PHOTO,
                status = MessageStatus.SENT
            )
            _messages.value = _messages.value + newMsg
            MessageRepository.addMessage(currentMatch.id, newMsg)

            if (currentMatch.firstMsgBy == null) {
                _match.value = currentMatch.copy(
                    firstMsgBy = "current-user-id",
                    firstMsgLocked = true,
                    firstMsgAt = System.currentTimeMillis()
                )
            }

            _isSending.value = false
            simulateReply(currentMatch.id)
        }
    }

    private fun receiveMessage(message: Message) {
        _messages.value = _messages.value + message
        MessageRepository.addMessage(message.matchId, message)
        val cm = _match.value ?: return

        // If we were waiting for a reply (firstMsgLocked), activate the match
        if (cm.firstMsgLocked || cm.status == MatchStatus.PENDING_FIRST_MESSAGE) {
            _match.value = cm.copy(
                firstMsgLocked = false,
                status = MatchStatus.ACTIVE,
                expiresAt = null,
                lastMessage = message.content
            )
            _chatUnlocked.value = true

            // Notify InboxViewModel — move match to Chats permanently
            onMatchActivated?.invoke(cm.id, message.content)

            viewModelScope.launch {
                delay(3000)
                _chatUnlocked.value = false
            }
        } else {
            // Normal message in an active chat — just update lastMessage
            _match.value = cm.copy(lastMessage = message.content)
        }
    }

    private fun checkAndUnlockIfReplied() {
        val cm = _match.value ?: return
        if ((cm.firstMsgLocked || cm.status == MatchStatus.PENDING_FIRST_MESSAGE) &&
            cm.firstMsgBy == "current-user-id" &&
            _messages.value.any { it.senderId != "current-user-id" }
        ) {
            val lastReply = _messages.value.lastOrNull { it.senderId != "current-user-id" }
            _match.value = cm.copy(
                firstMsgLocked = false,
                status = MatchStatus.ACTIVE,
                expiresAt = null,
                lastMessage = lastReply?.content
            )
            _chatUnlocked.value = true
            onMatchActivated?.invoke(cm.id, lastReply?.content ?: "")
            viewModelScope.launch {
                delay(3000)
                _chatUnlocked.value = false
            }
        }
    }

    private fun simulateReply(matchId: String) {
        viewModelScope.launch {
            delay(3000)
            _isOtherTyping.value = true
            delay(2000)
            _isOtherTyping.value = false
            val replies = listOf(
                "That's so sweet! Tell me more about yourself",
                "Haha I love that! What else are you into?",
                "Oh wow, we have so much in common!",
                "That's interesting! I'd love to hear more",
                "You seem really cool! What do you do for fun?"
            )
            val replyText = replies.random()
            receiveMessage(
                Message(
                    id = UUID.randomUUID().toString(),
                    matchId = matchId,
                    senderId = _match.value?.otherUser?.id ?: "other",
                    content = replyText,
                    status = MessageStatus.DELIVERED
                )
            )
        }
    }

    fun markUnreadMessagesAsRead() {
        _messages.value = _messages.value.map {
            if (!it.isFromMe && it.status != MessageStatus.READ) it.copy(status = MessageStatus.READ) else it
        }
    }
}
